Add initial_env param to FakeTensorProp

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96350
* __->__ #96879
* #96349

This PR allows users to skip nodes on FakeTensorProp.propagate.

titaiwangms committed:
Add override option to FakeTensorProp

[ghstack-poisoned]

wschin reviewed (approved):
LGTM.

wschin commented on torch/fx/passes/fake_tensor_prop.py:
```diff
@@ -20,16 +22,20 @@ class FakeTensorProp(torch.fx.Interpreter):
+             If False, only the Nodes without `val` will be assigned the fake tensor result.
     """
-    def __init__(self, module: torch.fx.GraphModule, mode: Optional[FakeTensorMode] = None):
+    def __init__(self, module: torch.fx.GraphModule, mode: Optional[FakeTensorMode] = None, override: bool = True):
```
```suggestion
    def __init__(self, module: torch.fx.GraphModule, mode: Optional[FakeTensorMode] = None, override: Optional[bool]= True):
```

justinchuby replied:
`Optional[bool]` would mean we expect None as an input. Was this intended?

titaiwangms replied:
I prefer this to be default True. Optional to None doesn't seen to be a good idea? Because I would like to keep it's default behavior.

titaiwangms committed:
Update on "Add override option to FakeTensorProp"


This PR allows users to use FakeTensorProp to only fill in meta value to those empty fx.Node.

[ghstack-poisoned]

eellison reviewed (commented):
The idiomatic way to stop a node from being run in interpreter is to pass in a value for it to `env` before running. Would that work here ? 

https://github.com/pytorch/pytorch/blob/master/torch/fx/interpreter.py#L105-L108

This is not a very clear api change IMO.

ezyang commented:
That seems awkward for shape propagation as is being done here though. I agree the change here is a bit unsavory but I'm not sure what a more principled fix looks like.

titaiwangms committed:
Update on "Add override option to FakeTensorProp"


This PR allows users to use FakeTensorProp to only fill in meta value to those empty fx.Node.

[ghstack-poisoned]

BowenBao commented:
@ezyang is there a symbolic version of this pass?

titaiwangms committed:
Update on "Add override option to FakeTensorProp"


This PR allows users to use FakeTensorProp to only fill in meta value to those empty fx.Node.

[ghstack-poisoned]

titaiwangms commented:
> The idiomatic way to stop a node from being run in interpreter is to pass in a value for it to `env` before running. Would that work here ?
> 
> https://github.com/pytorch/pytorch/blob/master/torch/fx/interpreter.py#L105-L108
> 
> This is not a very clear api change IMO.

Thanks! I tried initial_env:

```python
    initial_env = {
        node: node.meta["val"]
        for node in decomposed_module.graph.nodes
        if "val" in node.meta
    }
```

Didn't exclude the placeholder leading to an error of input numbers. Now that without placeholder skipped, it works great!

ezyang commented:
@BowenBao I don't know. I also don't think people should be using this pass; tracing gives you meta['val'] with fake tensors that have all the info you need, you really shouldn't have to call a pass like this.

titaiwangms commented:
Closing this PR as we should have enough node meta info from make_fx
